network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+use network::{
+    common::NetworkPublicKeys,
+    connectivity_manager::{ConnectivityRequest, DiscoverySource},
+};
+use once_cell::sync::Lazy;
+use std::{convert::TryFrom, time::Instant};
+
+/// Histogram of idle time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_IDLE_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_idle_duration_s",
+            "Histogram of idle time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Histogram of busy time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_BUSY_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_busy_duration_s",
+            "Histogram of busy time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+pub struct ConfigurationChangeListener {
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    role: RoleType,
+}
+
+fn extract_validator_updates(validator_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let validator_keys = validator_set.payload().to_vec();
+    // Collect the set of address updates.
+    let updates: Vec<ConnectivityRequest> = validator_keys
+        .into_iter()
+        .map(|validator| {
+            ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::OnChain,
+                validator.account_address().clone(),
+                vec![
+                    NetworkAddress::try_from(&validator.config().validator_network_address)
+                        .expect("WTF"),
+                ],
+            )
+        })
+        .collect();
+    // Collect the set of EligibleNodes
+    updates.push(ConnectivityRequest::UpdateEligibleNodes(
+        validator_keys
+            .into_iter()
+            .map(|validator| {
+                (
+                    *validator.account_address(),
+                    NetworkPublicKeys {
+                        identity_public_key: validator.network_identity_public_key(),
+                        signing_public_key: validator.network_signing_public_key().clone(),
+                    },
+                )
+            })
+            .collect(),
+    ));
+    updates
+}
+
+fn extract_full_node_updates(full_node_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let full_node_infos = full_node_set.payload().to_vec();
+
+    // Collect the set of updated addresses.
+    let updates: Vec<ConnectivityRequest> = full_node_infos
+        .into_iter()
+        .map(|full_node| {
+            ConnectivityRequest::UpdateAddresses(
+                DiscoverySource::OnChain,
+                // TODO:  this seems wrong for the account address.  Is it shared with the validator?
+                full_node.account_address().clone(),
```

Comment:
well this is truly a mess... we actually want this to be account_address::from_public_key where the public key is the identity key. We really need to land on how handle peers with inside connectivity manager otherwise we have a nice memory leak -- since that's currently a hash and not some cache limited capacity hash.